docs/source/core/core.rst

```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
+-----------------------
+
+When comparing two neural types the following comparison results can be generated.
+
+.. autoclass:: nemo.core.neural_types.NeuralTypeComparisonResult
+
+
+Examples
+--------
+
+Long vs short notation
+~~~~~~~~~~~~~~~~~~~~~~
+NeMo's NeuralType class allows you to express axis semantics information in long and short form.
+Consider these two equivalent types. Both encoder 3 dimensional tensors and both contain elements of type `AcousticEncodedRepresentation` (this type is a typical output of ASR encoders).
+
+.. code-block:: python
+
+    long_version = NeuralType(
+            axes=(AxisType(AxisKind.Batch, None), AxisType(AxisKind.Dimension, None), AxisType(AxisKind.Time, None)),
+            elements_type=AcousticEncodedRepresentation(),
+        )
+    short_version = NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation())
+    assert long_version.compare(short_version) == NeuralTypeComparisonResult.SAME
+
+
+
+Transpose same
+~~~~~~~~~~~~~~
+Often it is useful to know if a simple transposition will solve type incompatibility.
+This is the case if the comparison result of two types equals `nemo.core.neural_types.NeuralTypeComparisonResult.TRANSPOSE_SAME`.
+
+.. code-block:: python
+
+    type1 = NeuralType(axes=('B', 'T', 'C'))
+    type2 = NeuralType(axes=('T', 'B', 'C'))
+    assert type1.compare(type2) == NeuralTypeComparisonResult.TRANSPOSE_SAME
+    assert type2.compare(type1) == NeuralTypeComparisonResult.TRANSPOSE_SAME
+
+Note that in this example, we dropped `elements_type` argument of `NeuralType` constructor.
+If note supplied, the element type would be VoidType.
+
+VoidType for elements
+~~~~~~~~~~~~~~~~~~~~~
+Sometimes it is useful to express that elements' types don't matter but axes layout does.
+VoidType for elements can be used to express this.
+
+.. note::
+    VoidType is compatible with every other elements' type but not the other way around. See code snippet below for details.
+
+.. code-block:: python
+
+        btc_spctr = NeuralType(('B', 'T', 'C'), SpectrogramType())
+        btc_spct_bad = NeuralType(('B', 'T'), SpectrogramType())
+        # Note the VoidType for elements here
+        btc_void = NeuralType(('B', 'T', 'C'), VoidType())
+
+        # This is true because VoidType is compatible with every other element type (SpectrogramType in this case)
+        # And axes layout between btc_void and btc_spctr is the same
+        assert btc_void.compare(btc_spctr) == NeuralTypeComparisonResult.SAME
+        # These two types are incompatible because even though VoidType is used for elements on one side,
+        # the axes layout is different
+        assert btc_void.compare(btc_spct_bad) == NeuralTypeComparisonResult.INCOMPATIBLE
+        # Note that even though VoidType is compatible with every other type, other types are not compatible with VoidType!
+        # It is one-way compatibility
+        assert btc_spctr.compare(btc_void) == NeuralTypeComparisonResult.INCOMPATIBLE
```

Comment:
Possibly in the future we might want to make it symmetric. Not for RC though,

Reply:
no, this is intentionally non symmetric. Think of a function which accepts (void*)  --- it should accept every type. but (int) should only accept int and reject void.